README.md
django-zxcvbn-password
=======

A combination of
[piran’s django-zxcvbn](https://github.com/piran/django-zxcvbn) and
[aj-may’s django-password-strength](https://github.com/aj-may/django-password-strength)
Django apps.

It combines back-end and front-end validation with strength meter display.

Installation
------------

Run `pip install django-zxcvbn-password`  
Then in your settings:

```python
INSTALLED_APPS = (
    ...
    ‘django_zxcvbn_password’,
    ...
)
```

Usage
-----

```python
# forms.py

from django import forms
from django_zxcvbn_password.fields import PasswordField, PasswordConfirmationField

class RegisterForm(forms.Form):
    password1 = PasswordField()
    password2 = PasswordConfirmationField(confirm_with=’password1’)
```

Remember to include `{{ form.media }}` in your template.
Please refer to the documentation of the two upstream repository for more information.

Screenshot
----------

![Screenshot](http://img15.hostingpics.net/pics/295712Capturedu20150201153746.png)


Configuration
-------------

You can configure minimum and maximum length of the password,
and the minimum entropy (refer to zxcvbn documentation for more details).
The following values are the default:

```python
# settings.py

PASSWORD_MIN_LENGTH = 8
PASSWORD_MAX_LENGTH = 128
PASSWORD_MIN_ENTROPY = 25
```

License
-------

zxcvbn (https://github.com/lowe/zxcvbn),
django-zxcvbn (https://github.com/piran/django-zxcvbn) and
django-password-strength (https://github.com/aj-may/django-password-strength)
are redistributed under the terms of the BSD license, hence django-zxcvbn-password (https://github.com/Pawamoy/zxcvbn-password) is also redistributed under the terms of the BSD license:

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.



CHANGELOG.rst
=========
Changelog
=========

2.1.1 (2021-12-16)
==================

- Avoid using deprecated ugettext (PR #143).

2.1.0 (2019-12-15)
==================

- Allow specifying frequency lists in ``ZXCVBNValidator`` options (baa47cd).
- Return warnings as validationErrors, create list of warning/suggestion to return as ValidationError(s), fixing translations (12946bb).

2.0.3 (2019-02-21)
==================

- Use new location for package ``python-zxcvbn``, now ``zxcvbn`` (2ea1b69).


2.0.2 (2018-08-21)
==================

Documented
----------
- Improve usage notes (7a1ed42). Related issues/PRs: #31.

Fixed
-----
- Fix call to super in PasswordConfirmationInput (fc551b8).
- Improve password validator help text (c5d21a1). Related issues/PRs: #46.
- Strength bar color go green only when superior to min score (9a44fd8). Related issues/PRs: #3.

Tests
-----
- Add django 1.11 tests (815aaef).
- Add py37/pypy plus django 2.0 tests, remove py34 tests (05711cd).

2.0.1 (2017-02-17)
==================

* Fix call to super in PasswordStrengthInput.

2.0.0 (2017-02-17)
==================

* Drop Django 1.8 support in favor of AUTH_PASSWORD_VALIDATORS setting
  introduced in Django 1.9.
* Update zxcvbn to more recent version (dwolfhub/zxcvbn-python on GitHub).
* Update JavaScript code to latest version.
* Remove all settings (they now go in AUTH_PASSWORD_VALIDATOR options).
* Change license to ISC.

Thanks to Nick Stefan and Daniel Wolf.

1.1.0 (2016-10-18)
==================

* Cookiecutterize the project.

1.0.5 (2015-03-31)
==================

* I don't remember.

1.0.3 (2015-03-12)
==================

* Switch README to rst.
* Fix manifest rules.

1.0.2 (2015-03-12)
==================

* Change package name from django_zxcvbn_password to zxcvbn_password.

1.0.0 (2015-02-21)
==================

* Beta release on PyPI.

0.1.0 (2015-02-01)
==================

* Alpha release on PyPI.


validators.py
from django.core.exceptions import ValidationError
from django.utils.translation import ugettext_lazy as _
from django.conf import settings

import zxcvbn

# Settings
PASSWORD_MIN_LENGTH = getattr(settings, "PASSWORD_MIN_LENGTH", 8)
PASSWORD_MAX_LENGTH = getattr(settings, "PASSWORD_MAX_LENGTH", 128)
PASSWORD_MIN_ENTROPY = getattr(settings, "ZXCVBN_MIN_ENTROPY", 25)


class LengthValidator(object):
    code = "length"

    def __init__(self, min_length=None, max_length=None):
        self.min_length = min_length
        self.max_length = max_length

    def __call__(self, value):
        if self.min_length and len(value) < self.min_length:
            raise ValidationError(
                message=_("Must be %s characters or more.") % self.min_length,
                code=self.code)
        elif self.max_length and len(value) > self.max_length:
            raise ValidationError(
                message=_("Must be %s characters or less.") % self.max_length,
                code=self.code)


class ZXCVBNValidator(object):
    message = _("Try a bit more complex (rated at %s / %s)")
    code = "zxcvbn"

    def __call__(self, value):
        res = zxcvbn.password_strength(value)
        if res.get('entropy') < PASSWORD_MIN_ENTROPY:
            raise ValidationError(
                self.message % (res.get('entropy'), PASSWORD_MIN_ENTROPY),
                code=self.code)


length_validator = LengthValidator(PASSWORD_MIN_LENGTH, PASSWORD_MAX_LENGTH)
max_length_validator = LengthValidator(max_length=PASSWORD_MAX_LENGTH)
zxcvbn_validator = ZXCVBNValidator()